mongoengine/connection.py

```diff
@@ -398,6 +401,21 @@ def connect(db=None, alias=DEFAULT_CONNECTION_NAME, **kwargs):
     return get_connection(alias)
 
 
+def _set_session(session):
+    global _session
+    _session = session
```

Comment:
I would add a safeguard here and verify that `_session is None` when this is called, otherwise it would mean that we are doing something wrong

Reply:
This would also ensure that people don't start nesting the transaction by doing something like this:
(I'm actually not sure if pymongo would accept to instantiate the second session, let's also add a test for this case so we know what's the behavior) 

```
def do_something():
    with run_in_transaction():
        ...


with run_in_transaction():
    do_something()
```

Reply:
Fun fact - nested transactions are allowed apparently!

I've written two test cases to depict what happens when:
* the nested transactions throws an exc
* the parent transaction throws an exception AFTER the nested transaction was completed.

Reply:
Each time `run_in_transaction` was called - it instantiated a new session object and based on the tests, the rollback states work as I would expect them to.

I guess the question is do we want to force anything at the `mongoengine` level? My vote/opinion is: "if pymongo is OK with it, I'm OK with it"

Reply:
So turns out pymongo does in fact support nested sessions & transactions:
```python
from mongoengine import *
client = connect('mongoenginetest')

A = client.db.A
A.insert_one({"name": "a"})
B = client.db.B
B.insert_one({"name": "b"})


with client.start_session() as s1:
    with s1.start_transaction():
        A.update_one({"name": "a"}, {"$set": {"name": "txA"}}, session=s1)
        with client.start_session() as s2:
            with s2.start_transaction():
                B.update_one({"name": "b"}, {"$set": {"name": "txB"}}, session=s2)
                raise Exception

# Tx was rolled back
list(A.find({})), list(B.find({}))
# ([{'_id': ObjectId('618ae527af50dd19efc2344f'), 'name': 'a'}], [{'_id': ObjectId('618ae527af50dd19efc23450'), 'name': 'b'}])


with client.start_session() as s1:
    with s1.start_transaction():
        A.update_one({"name": "a"}, {"$set": {"name": "txA"}}, session=s1)
        with client.start_session() as s2:
            with s2.start_transaction():
                B.update_one({"name": "b"}, {"$set": {"name": "txB"}}, session=s2)
            raise Exception

# Inner tx was successful, outer tx was rolled back
list(A.find({})), list(B.find({}))
# ([{'_id': ObjectId('618ae527af50dd19efc2344f'), 'name': 'a'}], [{'_id': ObjectId('618ae527af50dd19efc23450'), 'name': 'txB'}])
```

I will start working on expanding the feature to support this. However, do let me know if you prefer to keep things simple for now, for example: "mongoengine will only support one session at a time and will add nested session support in the future"

This is an interesting one :)

Reply:
Turns out the nesting wasn't too tricky!